keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,196 @@
+* Continuing to support projects that import the credential provider package.
+
+## Proposal
+
+### External Credential Provider
+
+An executable capable of providing container registry credentials will be installed on each node. This binary will be executed by the kubelet to obtain container registry credentials in a format compatible with container runtimes. Credential responses may be cached within the kubelet.
+
+This architecture is similar to the approach taken by the exec based credential plugin architecture already present in client-go, and CNI and is a well understood pattern.  The API types are modeled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider configuration API type (similar to [clientauthentication](https://github.com/kubernetes/kubernetes/tree/0273d43ae9486e9d0be292c01de2dd4143522b86/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1))will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URIs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatch is used to match against the Image URL to determine which images to
+    // provide credentials for
+    ImageURLMatch string
+
+    // Exec specifies a custom exec-based authentication plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in JSON and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
+
+We will create new types `RegistryCredentialProviderPluginSpec` and `RegistryCredentialProviderPluginStatus` which will define the interface between the plugin and the kubelet runtime.  After the kubelet matches the image URL to a RegistryCredentialProvider, the kubelet will exec the plugin binary, and pass the JSON encoded spec to the plugin via stdin.  This includes the image URL that is to be pulled.  The plugin will report back the status, which includes the credentials that kubelet needs to pull the image.
+
+
+In the in-tree implementation, the docker keyring, which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.  This struct will be populated by the plugin rather than the in-tree provider.
+
+```go
+// RegistryCredentialProviderPluginSpec holds request and runtime specific information provided by
+// the transport.
+type RegistryCredentialProviderPluginSpec struct {
+    // +optional
+    Response *Response `json:"response,omitempty"`
```

Comment:
What's this?

Reply:
Whoops, that was supposed to be deleted, it was in the forked struct.